lib/CliftTransforms/ImmediateRadixDeduction.cpp

```diff
@@ -0,0 +1,125 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/APSInt.h"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "revng/Clift/Clift.h"
+#include "revng/CliftTransforms/Passes.h"
+
+namespace mlir {
+namespace clift {
+#define GEN_PASS_DEF_CLIFTIMMEDIATERADIXDEDUCTION
+#include "revng/CliftTransforms/Passes.h.inc"
+} // namespace clift
+} // namespace mlir
+
+namespace clift = mlir::clift;
+using namespace clift;
+
+namespace {
+
+// This multiplier describes a weighing of the entropy comparison in favour of
+// decimal (when the multiplier is greater than one).
+static constexpr double HexEntropyMultiplier = 1.20;
+
+template<uint64_t Radix>
+static double computeRepresentationEntropy(uint64_t Value) {
+  revng_assert(Value != 0);
+
+  uint8_t Digits[Radix] = {};
+  uint8_t UniqueDigitCount = 0;
+  uint8_t DigitCount = 0;
+
+  do {
+    uint8_t Digit = Value % Radix;
+    Value = Value / Radix;
+    UniqueDigitCount += Digits[Digit]++ == 0;
+    ++DigitCount;
+  } while (Value != 0);
+
+  return static_cast<double>(UniqueDigitCount) / DigitCount;
+}
+
+// Deduces the best radix based purely the value of an integer immediate.
+static unsigned deduceBestIntegerRadix(uint64_t Value) {
+  if (Value < 0x10)
+    return 10;
+
+  if (Value > 64 and std::has_single_bit(Value))
+    return 0x10;
+
+  double DecEntropy = computeRepresentationEntropy<10>(Value);
+  double HexEntropy = computeRepresentationEntropy<16>(Value);
+
+  if (HexEntropy * HexEntropyMultiplier < DecEntropy)
+    return 0x10;
+
+  return 10;
+}
+
+static bool isBitwiseOperator(mlir::Operation *Op) {
+  return mlir::isa<BitwiseNotOp, BitwiseAndOp, BitwiseOrOp, BitwiseXorOp>(Op);
+}
+
+// Deduces the best radix for an immediate based both on its usage and value.
+static unsigned selectIntegerRadix(ImmediateOp Immediate) {
+  for (mlir::Operation *User : Immediate.getResult().getUsers()) {
+    // Operands of bitwise operators are always displayed in hexadecimal:
+    if (isBitwiseOperator(User))
+      return 0x10;
+  }
+
+  return deduceBestIntegerRadix(Immediate.getValue());
+}
+
+// For each integer immediate, deduce the best radix and if it is not 10 (the
+// default), attach it in a discardable integer attribute named "clift.radix".
+struct ImmediateRadixDeductionPattern : mlir::OpRewritePattern<ImmediateOp> {
```

Comment:
This `RewritePattern` is not handling how we emit case labels.

I'd like to have an `OpRewritePatter<SwitchOp>` that applies the same heuristic at all the case labels, then picks the radix used by the majority of them and attaches a `clift.radix` attribute to the `SwitchOp` itself. The attribute is finally used by the backend to print the literals of each switch case in the same way.

The `RewritePattern`s for `ImmediateOp`s and `SwitchOp`s must be separate of course, but I'd like them both to run as part of the same pass.